File: 04-computation/continuing14_20260908_lrc_independent_audit.py
"""Independent full-domain and fully unpruned native-ratio LRC referee.

No primary producer import/execution. The C++ companion reconstructs all
geometry from raw strict walls and signed events, and exhausts every word.
"""
from pathlib import Path
from hashlib import sha256
from itertools import combinations
from collections import Counter,defaultdict
from math import gcd,comb
import argparse
import json
import os
import shutil
import subprocess
import sys
import tempfile

sys.stdout.reconfigure(encoding='utf-8',newline='\n')
HERE=Path(__file__).resolve().parent
STEM=Path(__file__).stem
PRIMARY='continuing14_20260908_lrc_zero_classification'
GATES=Counter()

def check(ok,label):
    GATES[label]+=1
    if not ok:raise ArithmeticError('always-active audit gate failed: '+label)

def canonical(value):return json.dumps(value,sort_keys=True,separators=(',',':')).encode()
def semantic(value):return sha256(canonical(value)).hexdigest()
def read_rows(path):return [list(map(int,line.split())) for line in path.read_text().splitlines()]

def connected_parts(word,table,t,component):
    parent=list(range(7))
    def root(i):
        while parent[i]!=i:i=parent[i]
        return i
    for i,j in combinations(range(7),2):
        e=gcd(word[i],word[j]);a,b=sorted((word[i]//e,word[j]//e))
        if table.get((t//e,a,b),(0,0,0))[component]:parent[root(i)]=root(j)
    groups=defaultdict(list)
    for i in range(7):groups[root(i)].append(i)
    return sorted(groups.values())

def ratio_count(table,t,a,b):
    e=gcd(a,b);aa,bb=sorted((a//e,b//e))
    return table.get((t//e,aa,bb),(0,0,0))[2]

def first_slot_failure(word,table,t):
    mult=Counter(word)
    for a in sorted(mult):
        if ratio_count(table,t,a,a):continue
        terms=[[b,mult[b],ratio_count(table,t,a,b)] for b in sorted(mult) if b!=a and ratio_count(table,t,a,b)]
        bound=sum(number*slots for b,number,slots in terms)
        if mult[a]>bound:return {'margin':a,'multiplicity':mult[a],'slot_bound':bound,'terms':terms}
    return None

def main():
    filed=HERE.name=='04-computation'
    parser=argparse.ArgumentParser()
    parser.add_argument('--root',type=Path,default=HERE.parent if filed else Path('C:/w/s0905'))
    parser.add_argument('--producer',type=Path,default=HERE.parent/'05-knowledge/results' if filed else Path('C:/w/continuing14_20260908_lrc'))
    args=parser.parse_args();root=args.root.resolve();producer=args.producer.resolve()
    expected_pins={'.py':'77b7edaa50e1a067a554ac330ee41d27b2aa1201797abd37b0091c8c636596e5',
        '.cpp':'dc7b5aec55e727963523cba0f976b9f9caaaa8c8d162ee4e4ad09b83b6a49aaf',
        '.out':'70b45b15b37f545f1d5af826abd6a1a2d77f4e65d9f8b8582a723a610b010b4c',
        '_certificate.json':'863d33accf19a5515b3ecd2048aac35d77f30f4e004371c73f5760d345bbb240'}
    for suffix,digest in expected_pins.items():
        directory=root/'04-computation' if filed and suffix in ('.py','.cpp') else producer
        check(sha256((directory/(PRIMARY+suffix)).read_bytes()).hexdigest()==digest,'frozen primary artifact '+suffix)
    primary=json.loads((producer/(PRIMARY+'_certificate.json')).read_bytes())
    raw=(root/'04-computation/overnight12_20260906_lrc_decoder_descent_inherited_profiles.json').read_bytes()
    check(sha256(raw).hexdigest()=='935f3f687b6d7c89cc099e536f536238fd753bcc4c1747906d213cef387ca93f','complete inherited profile pin')
    profiles=json.loads(raw)['levels'];alphabet=profiles['6']['gcds']
    check(alphabet==sorted({r[0] for r in profiles['6']['profiles']}) and len(alphabet)==42,'full original sheet alphabet')
    check(all(gcd(a,7)==1 for a in alphabet),'every sheet prime to seven')
    baseline_raw=(root/'05-knowledge/results/continuing13_20260908_lrc_zero_cuts_certificate.json').read_bytes()
    check(sha256(baseline_raw).hexdigest()=='23ae81ba6cb39bc14aecf2d93fc3fcf6d4e3c587b5240db35ea1d8f3992551f1','full predecessor certificate')
    baseline=json.loads(baseline_raw)['new_scales']
    check(len(baseline)==7618 and max(baseline)==11935 and semantic(baseline)=='89a3e7545cc77467c5f85fbe0bdaab1f71071cff65184bbc1212eb86c9f07177','complete predecessor necessary array')
    clocks=[t for t in baseline if t%7==0]
    check(len(clocks)==1085 and clocks==primary['declared_clocks'],'whole E0 survey selected before outcomes')
    domains=sorted({tuple(a for a in alphabet if t%a==0) for t in clocks});domain_index={d:i for i,d in enumerate(domains)}
    check(len(domains)==203 and [list(d) for d in domains]==[r['domain'] for r in primary['domains']],'every full divisor alphabet')
    quotients={t//gcd(a,b) for t in clocks for a in alphabet if t%a==0 for b in alphabet if t%b==0}
    check(quotients==set(clocks),'all actual pair quotient clocks covered exactly')
    profile_rows=[(int(k),c,word) for k,level in profiles.items() for c,word in level['profiles']]
    rows=[str(len(profile_rows))]+[' '.join(map(str,[k,c,*word])) for k,c,word in profile_rows]
    rows += [str(len(domains))]+[' '.join(map(str,[len(d),*d])) for d in domains]
    rows += [str(len(clocks))]+[f'{t} {domain_index[tuple(a for a in alphabet if t%a==0)]}' for t in clocks]
    temp_parent=Path(tempfile.gettempdir()).resolve()
    work=Path(tempfile.mkdtemp(prefix=STEM+'_',dir=temp_parent)).resolve()
    check(work.parent==temp_parent and work.name.startswith(STEM+'_'),'private native outputs confined to declared temporary parent')
    input_path=work/'input.txt';input_path.write_bytes(('\n'.join(rows)+'\n').encode())
    compiler=shutil.which('g++') or 'C:/Users/Eliott/scoop/apps/gcc/current/bin/g++.exe'
    executable=work/('audit.exe' if os.name=='nt' else 'audit')
    options=['-std=c++17','-O3','-DNDEBUG'] if sys.flags.optimize else ['-std=c++17','-O2']
    build=subprocess.run([compiler,*options,str(HERE/(STEM+'.cpp')),'-o',str(executable)],capture_output=True)
    check(build.returncode==0,'independent native compilation '+build.stderr.decode(errors='replace'))
    env=os.environ.copy();env['PATH']=str(Path(compiler).resolve().parent)+os.pathsep+env.get('PATH','')
    run=subprocess.run([str(executable),str(input_path),str(work)],capture_output=True,env=env)
    check(run.returncode==0,'independent native survey '+run.stderr.decode(errors='replace'))
    native=run.stdout.replace(b'\r\n',b'\n').decode()
    table_rows=read_rows(work/'zero_pairs.txt')
    check(table_rows==primary['quotient_zero_pairs'] and len(table_rows)==728,'complete unpruned native oriented ratio counts')
    table={tuple(r[:3]):tuple(r[3:]) for r in table_rows}
    raw_zero=read_rows(work/'all_zero_ratios.txt');ratio_sets=defaultdict(lambda:[set(),set()])
    for n,p,q,a,b,bits in raw_zero:
        check(gcd(p,q)==1 and p<q and gcd(n,p)==a and gcd(n,q)==b,'every individual native ratio typed')
        for left,right,numerator,denominator in ((a,b,p,q),(b,a,q,p)):
            ratio_sets[n,left,right][0].add((numerator,denominator))
            if bits&2:ratio_sets[n,left,right][1].add((numerator,denominator))
    for n,a,b,bits,c,o in table_rows:
        check(len(ratio_sets[n,a,b][0])==c and len(ratio_sets[n,a,b][1])==o,'distinct oriented sets reconstruct every slot count')
        check(ratio_sets[n,a,b][1]<=ratio_sets[n,a,b][0],'complete chamber ratio subset')
    critical=read_rows(work/'critical_banks.txt')
    check(critical==primary['critical_native_banks'],'every retained critical native bank from unpruned atlas')
    domain_rows=read_rows(work/'domains.txt');wordbanks=[];raw_count=0;valid_count=0
    for i,count,raw_multisets in domain_rows:
        data=(work/f'words_{i}.json').read_bytes();wordbanks.append(json.loads(data))
        expected=primary['domains'][i]
        check(len(wordbanks[-1])==count==expected['word_count'],'entire independently accepted word-bank length')
        check(raw_multisets==comb(len(domains[i])+6,7)==expected['unpruned_multisets'],'unpruned complete multiset universe')
        check(sha256(data).hexdigest()==expected['words_sha256'],'whole independently regenerated word-bank digest')
        raw_count+=raw_multisets;valid_count+=count
    check(raw_count==9978787 and valid_count==415506,'full exact domain census totals')
    records=read_rows(work/'survey.txt')
    check(records==primary['clocks'],'all 1085 independent clock records, stages and first owners')
    check(sum(r[2] for r in records)==630818,'all complete word-clock evaluations')
    stages=[[r[0] for r in records if r[3]==0],
            [r[0] for r in records if r[3]>0 and r[4]==0],
            [r[0] for r in records if r[4]>0 and r[5]==0]]
    check(list(map(len,stages))==[713,8,8],'three disjoint deletion stages')
    deleted=set();intermediate=[]
    for actual,expected in zip(stages,primary['stages']):
        check(actual==expected['removed_clocks'] and not deleted.intersection(actual),'stage has only its independently proved new removals')
        deleted.update(actual);remaining=[t for t in baseline if t not in deleted]
        check(remaining==expected['new_scales'] and semantic(remaining)==expected['new_scales_sha256'],'complete intermediate array and semantic pin')
        intermediate.append(len(remaining))
    check(intermediate==[6905,6897,6889],'all intermediate counts')
    check(max(remaining)==11934 and semantic(remaining)=='b603ca0186c1ebaa2318782348a7446a16f529543147eb53062f1ba360e8e57b','complete final array and maximum')
    check(remaining==primary['new_scales'],'full final necessary array matches certificate')
    e0=[t for t in remaining if t%7==0]
    check(e0==primary['remaining_E0'] and len(e0)==356 and max(e0)==7056,'whole residual E0 slice')
    refinement_pairs=0
    for j,(T,d,*_) in enumerate(records):
        for t,small_d,*_ in records[:j]:
            if small_d!=d or T%t:continue
            refinement_pairs+=1
            for a in domains[d]:
                for b in domains[d]:
                    e=gcd(a,b)
                    small=ratio_sets[t//e,a//e,b//e];large=ratio_sets[T//e,a//e,b//e]
                    check(large[0]<=small[0] and large[1]<=small[1],'full oriented-set inclusion for every lawful divisor refinement')
    check(refinement_pairs==primary['lawful_divisible_refinement_pairs']==1055,'complete lawful refinement universe')
    refined=[]
    for expected in primary['refined_complete_residuals']:
        t=expected['clock'];d=domain_index[tuple(a for a in alphabet if t%a==0)]
        closed=[];opened=[];slots=[];proofs=[]
        for word in wordbanks[d]:
            parts0=connected_parts(word,table,t,1);parts1=connected_parts(word,table,t,2)
            if len(parts0)==1:closed.append(word)
            if len(parts1)>1:
                if len(parts0)==1:proofs.append({'word':word,'chamber_components':parts1})
                continue
            opened.append(word);failure=first_slot_failure(word,table,t)
            if failure:proofs.append({'word':word,'slot_failure':failure})
            else:slots.append(word)
        actual={'clock':t,'closed_residuals':closed,'chamber_residuals':opened,'slot_residuals':slots,'proofs':proofs}
        check(actual==expected,'entire refined residual bank and every explicit cut or slot witness')
        refined.append(actual)
    stop=read_rows(work/'stop.txt')
    check(stop==[[4,8,8,9,9,18,24],[4,8,9,9,16,18,24],[4,9,9,16,16,18,24]],'complete 7056 stopping words')
    check(len([w for r in refined if r['clock'] in stages[1] for w in r['closed_residuals']])==18,'all 18 ordinary-only refined words')
    check(len([w for r in refined if r['clock'] in stages[2] for w in r['chamber_residuals']])==31,'all 31 slot-only refined words')
    native_gates=int(next(line.split()[-1] for line in native.splitlines() if line.startswith('NATIVE_ALWAYS_ACTIVE_GATES ')))
    certificate={'status':'INDEPENDENTLY ACCEPTED complete scoped E0 classification',
        'reviewed_primary_report_sha256':'0ae9aa1b4d52989b69da202b8981ac41f39075cda8da90d557668512cbd0b1d3',
        'primary_executable_pins':expected_pins,'native_audit_source_sha256':sha256((HERE/(STEM+'.cpp')).read_bytes()).hexdigest(),
        'native_transcript':native,'native_gates':native_gates,'python_gate_counts':dict(sorted(GATES.items())),
        'python_gates':sum(GATES.values()),'total_gates':native_gates+sum(GATES.values()),
        'complete_domains':203,'unpruned_multisets':raw_count,'distinct_valid_words':valid_count,
        'word_clock_evaluations':630818,'all_zero_ratio_rows':len(raw_zero),
        'all_zero_ratios_sha256':sha256((work/'all_zero_ratios.txt').read_bytes()).hexdigest(),
        'quotient_zero_pairs':table_rows,'separate_removed_clocks':stages,'intermediate_counts':intermediate,
        'new_scales':remaining,'new_scales_sha256':semantic(remaining),'remaining_E0':e0,
        'lawful_full_ratio_set_refinements':refinement_pairs,'stop_words':stop,
        'scope':'Distinct positive speeds, selected-six gcd clock, actual connected strict complement. Remaining graphs are not joint ratio/phase realizations.'}
    destination=root/'05-knowledge/results' if filed else HERE
    (destination/(STEM+'_certificate.json')).write_bytes(canonical(certificate)+b'\n')
    print('Independent complete zero-budget clock classification: PASS')
    print(native.rstrip())
    print('All 203 word banks, 728 complete oriented zero-pair rows, 1085 clock records, and 1055 full ratio-set refinements agree')
    print('SEPARATE_REMOVALS 713+8+8; NECESSARY_ARRAY 6889 MAX 11934; E0_ARRAY 356 MAX 7056')
    print('FINAL_ARRAY_SHA256',semantic(remaining))
    print('PYTHON_ALWAYS_ACTIVE_GATES',sum(GATES.values()))
    print('TOTAL_ALWAYS_ACTIVE_GATES',native_gates+sum(GATES.values()))

if __name__=='__main__':main()

File: 04-computation/continuing14_20260908_lrc_independent_audit.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <map>
#include <numeric>
#include <set>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;
using I=long long;
using Word=array<int,7>;
uint64_t gates=0,raw_words=0,capacity_evaluations=0,zero_ratios=0,open_zero_ratios=0,literal_points=0;
void need(bool ok,const string& why){++gates;if(!ok){cerr<<"FAILED "<<why<<'\n';exit(1);}}
I mod(I a,I b){I r=a%b;return r<0?r+b:r;}
I floorq(I a,I b){return (a-mod(a,b))/b;}
I ceilq(I a,I b){return -floorq(-a,b);}
bool danger(I x,I denominator,int speed){I r=mod(speed*x,denominator);return 14*min(r,denominator-r)<denominator;}
struct Geometry{int p,q;I denominator;vector<pair<I,I>> intervals;};
Geometry geometry(int p,int q){
    Geometry G{p,q,14LL*p*q,{}};
    vector<I> boundaries={0,G.denominator};
    for(int j=0;j<=p;j++)for(int sign:{-1,1}){I x=(14LL*j+sign)*q;if(0<x&&x<G.denominator)boundaries.push_back(x);}
    for(int j=0;j<=q;j++)for(int sign:{-1,1}){I x=(14LL*j+sign)*p;if(0<x&&x<G.denominator)boundaries.push_back(x);}
    sort(boundaries.begin(),boundaries.end());boundaries.erase(unique(boundaries.begin(),boundaries.end()),boundaries.end());
    vector<pair<I,I>> cells;
    for(size_t j=1;j<boundaries.size();j++){
        I a=boundaries[j-1],b=boundaries[j];
        if(danger(a+b,2*G.denominator,p)&&danger(a+b,2*G.denominator,q)){
            if(!cells.empty()&&cells.back().second==a&&danger(a,G.denominator,p)&&danger(a,G.denominator,q))cells.back().second=b;
            else cells.push_back({a,b});
        }
    }
    need(cells.size()>=2&&cells.front().first==0&&cells.back().second==G.denominator,"raw midpoint geometry retains circle wrap");
    G.intervals.push_back({cells.back().first-G.denominator,cells.front().second});
    G.intervals.insert(G.intervals.end(),cells.begin()+1,cells.end()-1);
    return G;
}
I interval_count(const Geometry&G,int n,I phase2){
    I value=0;
    for(auto[a,b]:G.intervals)value+=ceilq(2LL*n*b-phase2,2*G.denominator)-floorq(2LL*n*a-phase2,2*G.denominator)-1;
    return value;
}
struct Capacity{I ordinary,chamber,ordinary_phase,chamber_phase;};
Capacity capacity(const Geometry&G,int n){
    vector<pair<I,int>> events;
    for(auto[a,b]:G.intervals){events.push_back({mod(n*a,G.denominator),1});events.push_back({mod(n*b,G.denominator),-1});}
    sort(events.begin(),events.end());
    I initial=interval_count(G,n,0),cur=interval_count(G,n,1);
    Capacity out{min(initial,cur),cur,initial<=cur?0:1,1};
    size_t j=0;
    while(j<events.size()){
        I wall=events[j].first;int starts=0,ends=0;
        while(j<events.size()&&events[j].first==wall){if(events[j].second==1)starts++;else ends++;j++;}
        if(wall==0)continue;
        I at=cur-ends;
        if(at<out.ordinary){out.ordinary=at;out.ordinary_phase=2*wall;}
        cur=at+starts;
        if(cur<out.ordinary){out.ordinary=cur;out.ordinary_phase=2*wall+1;}
        if(cur<out.chamber){out.chamber=cur;out.chamber_phase=2*wall+1;}
    }
    return out;
}
array<I,3> literal(const Geometry&G,int n,I phase2){
    array<I,3> counts{};I denominator=2*G.denominator*n;
    for(int j=0;j<n;j++){
        I x=2*G.denominator*j+phase2;bool a=danger(x,denominator,G.p),b=danger(x,denominator,G.q);
        counts[0]+=a;counts[1]+=b;counts[2]+=a&&b;literal_points++;
    }
    return counts;
}
vector<Geometry> atlas(){
    vector<int> primes;
    for(int p=2;p<=356;p++){
        bool prime=true;for(int d=2;d*d<=p;d++)if(p%d==0){prime=false;break;}
        if(prime&&p%3==2)primes.push_back(p);
    }
    set<int> sums={1};
    for(int p:primes){vector<int> before(sums.begin(),sums.end());for(int v:before){if(v*p<=356)sums.insert(v*p);if((I)v*p*p<=356)sums.insert(v*p*p);}}
    vector<Geometry> out;
    for(int v:sums)if(v>=3)for(int p=1;2*p<v;p++)if(gcd(p,v-p)==1)out.push_back(geometry(p,v-p));
    need(out.size()==5855,"independently multiplicative full strict atlas");return out;
}
void controls(){
    for(auto[p,q]:vector<pair<int,int>>{{1,1},{1,13},{1,112},{1,355},{11,263},{23,323},{33,320},{48,307}}){
        auto G=geometry(p,q);
        for(int n:{7,14,29,651,1260,1890,1939,1953,2485}){
            set<I> walls={0};for(auto[a,b]:G.intervals){walls.insert(mod(n*a,G.denominator));walls.insert(mod(n*b,G.denominator));}
            I ordinary=1000000000,chamber=1000000000;
            for(I w:walls){ordinary=min(ordinary,interval_count(G,n,2*w));I value=interval_count(G,n,2*w+1);ordinary=min(ordinary,value);chamber=min(chamber,value);}
            auto found=capacity(G,n);
            need(found.ordinary==ordinary&&found.chamber==chamber,"every direct wall and chamber versus independent signed sweep");
            need(literal(G,n,found.ordinary_phase)[2]==ordinary,"literal control native intersection");
        }
    }
}
int common[91][91];
array<unordered_set<uint64_t>,7> profiles;
vector<int> inside_masks;
bool valid(const Word&w){
    int total=0;for(int x:w)total=common[total][x];if(total!=1)return false;
    int gcd_subset[128]={};
    for(int mask=1;mask<128;mask++){
        int index=__builtin_ctz((unsigned)mask);gcd_subset[mask]=common[gcd_subset[mask&(mask-1)]][w[index]];
    }
    for(int selected:inside_masks){
        int c=gcd_subset[127^selected],k=__builtin_popcount((unsigned)selected),j=0;
        array<int,6> restricted{};
        for(int i=0;i<7;i++)if(selected&(1<<i))restricted[j++]=common[c][w[i]];
        sort(restricted.begin(),restricted.begin()+k);uint64_t signature=c;
        for(int i=0;i<k;i++)signature=(signature<<7)|restricted[i];
        if(!profiles[k].count(signature))return false;
    }
    return true;
}
void enumerate(const vector<int>& alphabet,int position,int first,Word&w,vector<Word>&bank){
    if(position==7){raw_words++;if(valid(w))bank.push_back(w);return;}
    for(int i=first;i<(int)alphabet.size();i++){w[position]=alphabet[i];enumerate(alphabet,position+1,i,w,bank);}
}
using Table=array<array<int,91>,91>;
bool connected(const Word&w,const Table& counts){
    array<int,7> parent; iota(parent.begin(),parent.end(),0);
    auto find=[&](int a){while(parent[a]!=a)a=parent[a];return a;};
    for(int i=0;i<7;i++)for(int j=i+1;j<7;j++)if(counts[w[i]][w[j]]>0)parent[find(i)]=find(j);
    for(int i=1;i<7;i++)if(find(i)!=find(0))return false;return true;
}
void writeword(ostream&out,const Word&w){for(int x:w)out<<' '<<x;}
int main(int argc,char**argv){
    need(argc==3,"independent input and output paths");ifstream input(argv[1]);string directory=argv[2];
    for(int a=0;a<=90;a++)for(int b=0;b<=90;b++)common[a][b]=gcd(a,b);
    int count;input>>count;
    for(int j=0;j<count;j++){int k,c;input>>k>>c;uint64_t signature=c;for(int i=0;i<k;i++){int x;input>>x;signature=(signature<<7)|x;}profiles[k].insert(signature);}
    for(int k=1;k<=6;k++)for(int mask=1;mask<127;mask++)if(__builtin_popcount((unsigned)mask)==k)inside_masks.push_back(mask);
    need(inside_masks.size()==126,"all proper selected-position profile tests");
    int nd;input>>nd;vector<vector<int>> alphabets(nd);
    for(auto&row:alphabets){int size;input>>size;row.resize(size);for(int&x:row)input>>x;}
    int nt;input>>nt;vector<pair<int,int>> clocks(nt);set<int> quotients,declared;
    for(auto&[t,d]:clocks){input>>t>>d;declared.insert(t);for(int a:alphabets[d])for(int b:alphabets[d]){need(t%(7*a)==0,"whole original margin alphabet zero budget");quotients.insert(t/gcd(a,b));}}
    need(bool(input)&&declared==quotients,"all actual quotient clocks covered by exact declared set");
    controls();auto ratios=atlas();
    map<int,pair<Table,Table>> banks;
    ofstream table_file(directory+"/zero_pairs.txt",ios::binary),critical_file(directory+"/critical_banks.txt",ios::binary),all_zero_file(directory+"/all_zero_ratios.txt",ios::binary);
    set<int> critical={1939,1953};for(int a:{4,8,9,16,18,24})for(int b:{4,8,9,16,18,24})critical.insert(7056/gcd(a,b));
    uint64_t examined=0;
    for(int n:quotients){
        Table ordinary{},chamber{};
        for(const auto&G:ratios){
            int a=gcd(n,G.p),b=gcd(n,G.q);examined++;
            if(a>90||b>90||a%7==0||b%7==0)continue;
            // Deliberately no lower-bound skip and no first-witness/ratio exit.
            auto value=capacity(G,n);capacity_evaluations++;
            need(value.ordinary>=0&&value.chamber>=value.ordinary,"unpruned native capacities ordered");
            int copies=a==b?2:1,lo=min(a,b),hi=max(a,b);
            if(value.ordinary==0){
                ordinary[lo][hi]+=copies;zero_ratios++;
                need(literal(G,n,value.ordinary_phase)[2]==0,"every ordinary zero has literal full-grid witness");
            }
            if(value.chamber==0){
                chamber[lo][hi]+=copies;open_zero_ratios++;
                auto physical=literal(G,n,value.chamber_phase);
                need(physical[2]==0&&physical[0]==n/7&&physical[1]==n/7,"every chamber zero attains both actual marginal bounds");
            }
            int bits=(value.ordinary==0?1:0)|(value.chamber==0?2:0);
            if(bits)all_zero_file<<n<<' '<<G.p<<' '<<G.q<<' '<<a<<' '<<b<<' '<<bits<<'\n';
            if(bits&&critical.count(n))critical_file<<n<<' '<<G.p<<' '<<G.q<<' '<<a<<' '<<b<<' '<<bits<<'\n';
        }
        for(int a=1;a<=90;a++)for(int b=a;b<=90;b++){
            if(ordinary[a][b])table_file<<n<<' '<<a<<' '<<b<<' '<<(1|(chamber[a][b]?2:0))<<' '<<ordinary[a][b]<<' '<<chamber[a][b]<<'\n';
            ordinary[b][a]=ordinary[a][b];chamber[b][a]=chamber[a][b];
        }
        banks.emplace(n,make_pair(ordinary,chamber));
    }
    cerr<<"UNPRUNED_NATIVE ratios="<<examined<<" typed="<<capacity_evaluations<<" zeros="<<zero_ratios<<'\n';
    vector<vector<Word>> words(nd);ofstream domain_file(directory+"/domains.txt",ios::binary);
    uint64_t valid_words=0;
    for(int d=0;d<nd;d++){
        uint64_t before=raw_words;Word w{};enumerate(alphabets[d],0,0,w,words[d]);valid_words+=words[d].size();
        domain_file<<d<<' '<<words[d].size()<<' '<<raw_words-before<<'\n';
        ofstream out(directory+"/words_"+to_string(d)+".json",ios::binary);out<<'[';bool first=true;
        for(const auto&row:words[d]){if(!first)out<<',';first=false;out<<'[';for(int j=0;j<7;j++){if(j)out<<',';out<<row[j];}out<<']';}out<<']';
    }
    cerr<<"UNPRUNED_DOMAINS raw="<<raw_words<<" valid="<<valid_words<<'\n';
    ofstream survey_file(directory+"/survey.txt",ios::binary),stop_file(directory+"/stop.txt",ios::binary);
    I evaluated_words=0,closed0=0,closed1=0,closed2=0;
    for(auto[t,d]:clocks){
        Table ordinary{},chamber{};
        for(int a:alphabets[d])for(int b:alphabets[d]){int e=gcd(a,b);ordinary[a][b]=banks.at(t/e).first[a/e][b/e];chamber[a][b]=banks.at(t/e).second[a/e][b/e];}
        I n0=0,n1=0,n2=0;Word first0{},first1{},first2{};
        for(const auto&w:words[d]){
            evaluated_words++;
            if(connected(w,ordinary)){if(!n0)first0=w;n0++;}
            if(!connected(w,chamber))continue;
            if(!n1)first1=w;n1++;
            map<int,int> multiplicity;for(int a:w)multiplicity[a]++;
            bool possible=true;
            for(auto[a,needed]:multiplicity)if(chamber[a][a]==0){
                I slots=0;for(auto[b,number]:multiplicity)if(a!=b)slots+=(I)number*chamber[a][b];
                if(slots<needed)possible=false;
            }
            if(possible){if(!n2)first2=w;n2++;if(t==7056){for(int a:w)stop_file<<a<<' ';stop_file<<'\n';}}
        }
        need(n2<=n1&&n1<=n0,"all independent clock-stage residuals nested");
        closed0+=n0==0;closed1+=n1==0;closed2+=n2==0;
        survey_file<<t<<' '<<d<<' '<<words[d].size()<<' '<<n0<<' '<<n1<<' '<<n2;
        writeword(survey_file,first0);writeword(survey_file,first1);writeword(survey_file,first2);survey_file<<'\n';
    }
    cout<<"ATLAS 5855; QUOTIENTS "<<quotients.size()<<"; DOMAINS "<<nd<<'\n';
    cout<<"NATIVE_RATIOS_EXAMINED "<<examined<<"; TYPED_CAPACITIES_NO_SKIPS "<<capacity_evaluations<<'\n';
    cout<<"ORDINARY_ZERO_RATIOS "<<zero_ratios<<"; CHAMBER_ZERO_RATIOS "<<open_zero_ratios<<"; LITERAL_GRID_POINTS "<<literal_points<<'\n';
    cout<<"UNPRUNED_MULTISETS "<<raw_words<<"; DISTINCT_VALID_WORDS "<<valid_words<<"; WORD_CLOCKS "<<evaluated_words<<'\n';
    cout<<"CUMULATIVE_CLOSURES "<<closed0<<' '<<closed1<<' '<<closed2<<'\n';
    cout<<"NATIVE_ALWAYS_ACTIVE_GATES "<<gates<<'\n';
}
